carbon-design-system/carbon
# feat(datepicker): add `parseDate` prop, improve default date parsing

Closes #15432 
Part of #13828

This improves DatePicker's date parsing functionality when using the default `dateFormat`. It also adds a new `parseDate` prop for consumers to supply their own date parsing function when using other date formats.

How this works:
* When using the default dateFormat, the DatePicker will attempt to parse the year, month and day into a Date object.
* If an invalid month or day is provided, `1` will be used instead. 
* If a totally invalid date is provided (`a1/0a/a999`), the input will be cleared.
* If a custom `dateFormat` is used and a `parseDate` is not provided, the default within flatpickr will be used, which can cause odd date correction based on how native date methods work, https://github.com/carbon-design-system/carbon/issues/15432#issuecomment-1967447677


#### Changelog

**New**

- Add `parseDate` prop to DatePicker
- Provide improved date parsing when using the default date format
- Add tests to cover the new default parsing, the passing of a `parseDate` prop function, and completely invalid dates.

#### Testing / Reviewing

- Go to the DatePicker stories and start entering various invalid dates. Month/day should correct to `01` and totally invalid dates should clear the input.
- Ensure it works on all DatePicker variants/stories (range, simple, etc)
